utilities.js

```diff
@@ -1632,6 +1632,102 @@ var Utilities = {
 	resolveIntermediateURL: movedToUtilitiesInternal("resolveIntermediateURL"),
 	determineAttachmentIcon: movedToUtilitiesInternal('determineAttachmentIcon'),
 
+	/**
+	 * Walk the DOM and the contents of JSON data attributes in the HTML representation
+	 * of a note, calling visitor functions that might modify it and returning
+	 * the resulting HTML.
+	 *
+	 * Elements are visited in depth-first order. First the element itself is visited,
+	 * then its data attributes, then the URIs in its JSON attributes, then its subtree.
+	 *
+	 * @param {String} note Note HTML
+	 * @param {Object} visitors
+	 * @param {Function} [visitors.visitContainer]
+	 * @param {Function} [visitors.visitAnnotation]
+	 * @param {Function} [visitors.visitCitation]
+	 * @param {Function} [visitors.visitOtherElement]
+	 * @param {Function} [visitors.visitDataAttribute]
+	 * @param {Function} [visitors.visitURI] Return a replacement for the passed URI
+	 * @return {String} Potentially modified note HTML
+	 */
+	walkNoteDOM: function (note, visitors) {
+		function visit(elem) {
+			if (elem.hasAttribute('data-schema-version')) {
+				visitors.visitContainer?.(elem);
+			}
+			else if (elem.hasAttribute('data-annotation')) {
+				visitors.visitAnnotation?.(elem);
+			}
+			else if (elem.hasAttribute('data-citation')) {
+				visitors.visitCitation?.(elem);
+			}
+			else {
+				visitors.visitOtherElement?.(elem);
+			}
+
+			if (visitors.visitDataAttribute || visitors.visitURI) {
+				for (let attr of ['data-citation', 'data-citation-items', 'data-annotation']) {
+					if (elem.hasAttribute(attr)) {
+						let json;
+						try {
+							json = JSON.parse(decodeURIComponent(elem.getAttribute(attr)));
+						}
+						catch (e) {
+							continue;
+						}
+
+						visitors.visitDataAttribute?.(attr, json);
+						if (visitors.visitURI) visitURIs(json);
+
+						elem.setAttribute(attr, JSON.stringify(json));
+					}
+				}
+			}
+
+			for (let child of elem.children) {
+				visit(child);
+			}
+		}
+
+		function visitURIs(json) {
+			if (Array.isArray(json)) {
+				json.forEach(visitURIs);
+			}
+			else if (typeof json === 'object') {
+				for (let [key, value] of Object.entries(json)) {
+					if (key == 'id' || key == 'attachmentURI') {
+						json[key] = visitors.visitURI(value);
+					}
+					else if (key == 'uris') {
+						json[key] = value.map(visitors.visitURI)
+					}
+					else {
+						visitURIs(value);
+					}
+				}
+			}
+		}
+
+		let wrappedNote = '<div id="note-body">' + note + '</div>';
+		let doc;
+		if (Zotero.isNode) {
+			let { JSDOM } = require('jsdom');
+			doc = new JSDOM(wrappedNote).window.document;
+		}
+		else {
+			let parser = typeof Cc !== 'undefined'
+				? Cc['@mozilla.org/xmlextras/domparser;1'].createInstance(Ci.nsIDOMParser)
+				: new DOMParser();
```

Comment:
Updated Firefox also uses `new DOMParser()`, so let's reverse this and use `typeof DOMParser`.

Reply:
For some reason, Fx60 does _define_ DOMParser in non-content contexts, but it just throws an `NS_ERROR_UNEXPECTED` when you try to construct it. So I guess we need a try/catch here until we're fully on Fx100.